jbmc/src/java_bytecode/ci_lazy_methods.cpp

```diff
@@ -438,15 +438,17 @@ void ci_lazy_methodst::initialize_instantiated_classes(
 ///   e that calls a virtual function.
 void ci_lazy_methodst::gather_virtual_callsites(
   const exprt &e,
-  std::unordered_set<exprt, irep_hash> &result)
+  std::unordered_set<virtual_function_exprt, irep_hash> &result)
 {
   if(e.id()!=ID_code)
     return;
   const codet &c=to_code(e);
-  if(c.get_statement()==ID_function_call &&
-     to_code_function_call(c).function().id()==ID_virtual_function)
+  if(
+    c.get_statement() == ID_function_call &&
+    can_cast_expr<virtual_function_exprt>(to_code_function_call(c).function()))
   {
-    result.insert(to_code_function_call(c).function());
+    result.insert(
+      to_virtual_function_expr(to_code_function_call(c).function()));
```

Comment:
consider `try_dynamic_cast_thing` (and test for `null` instead of `can_cast_expr`).

Reply:
It's too complicated to keep the current control flow to the else branch.